File: SolvdQuestions/76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        
        int m = s.length();
        int n = t.length();

        int remLen = n;

        // put all the chars of string t to map
        Map<Character, Integer> mp = new HashMap<>();

        for (char ch: t.toCharArray()) {
            mp.put(ch, mp.getOrDefault(ch, 0) + 1);
        }
        
        int startIndex = 0;
        int minLen = Integer.MAX_VALUE;
        int currLen = 0;

        int i=0, j=0;
        while (i < m) {
            char ch = s.charAt(i);
            if ( mp.containsKey(ch) ) {
                int val = mp.get(ch);
                if (val >= 1) {
                    remLen--;
                }
                mp.put(ch, val-1);
            }
            currLen++;
            while (remLen == 0) {
                char ch2 = s.charAt(j);
                if (mp.containsKey(ch2)) {
                    int val = mp.get(ch2);
                    if (val == 0) {
                        remLen++;
                    }
                    mp.put(ch2, mp.get(ch2) + 1);
                }
                if (minLen > i-j+1) {
                    minLen = i-j+1;
                    startIndex = j;
                }
                j++;
            }
            i++;
        }

        if (minLen == Integer.MAX_VALUE)
            return "";
        
        return s.substring(startIndex, startIndex+minLen);
    }
}

File: SolvdQuestions/76-minimum-window-substring/minimum-window-substring.go
// func min(a , b int) int {
//     if a < b {
//         return a
//     }
//     return b
// }


func minWindow(s string, t string) string {
    
    count := len(t)

    mp := make(map[byte]int)

    for i:=0; i<len(t); i++ {
        ch := t[i]
        if _, exists := mp[ch]; exists {
            mp[ch] += 1
        } else {
            mp[ch] = 1
        }
    }

    startIndex := 0
    minLen := math.MaxInt

    i := 0
    j := 0

    for j<len(s) {

        if _, exists := mp[s[j]]; exists {

            if mp[s[j]] > 0 {
                count--;
            }
            mp[s[j]] -= 1

            for (count == 0) {

                if j-i+1 < minLen {
                    minLen = j-i+1
                    startIndex = i
                }

                if _, exists := mp[s[i]]; exists {
                    mp[s[i]] += 1
                    if (mp[s[i]] >= 1) {
                        count++
                    }
                }
                
                i++
            }
        }
        j++
    }

    result := ""
    if minLen != math.MaxInt {
        result = s[startIndex : startIndex+minLen]
    }

    return result
}